spec/unit/lib/cloud_controller/deployment_updater/actions/up_scaler_spec.rb

```diff
@@ -0,0 +1,196 @@
+        state: DeploymentModel::DEPLOYING_STATE,
+        max_in_flight: max_in_flight
+      )
+    end
+
+    let(:logger) { instance_double(Steno::Logger, info: nil, error: nil) }
+    let(:current_instance_count) { 4 }
+    let(:max_in_flight) { 2 }
+
+    describe '#can_scale?' do
+      let(:current_instance_count) { 4 }
+      let(:max_in_flight) { 2 }
+
+      it 'returns true if it has enough max-in-flight to scale' do
+        summary = double(starting_instances_count: 1, routable_instances_count: 5, healthy_instances_count: nil, unhealthy_instances_count: 0)
+        up_scaler = DeploymentUpdater::Actions::UpScaler.new(deployment, logger, 10, summary)
+        expect(up_scaler.can_scale?).to be true
+      end
+
+      it 'returns false if there are any unhealthy instances' do
+        summary = double(starting_instances_count: 1, routable_instances_count: 5, healthy_instances_count: nil, unhealthy_instances_count: 1)
+        up_scaler = DeploymentUpdater::Actions::UpScaler.new(deployment, logger, 10, summary)
+        expect(up_scaler.can_scale?).to be false
+      end
+
+      it 'returns false if starting instances >= max_in_flight' do
+        summary = double(starting_instances_count: 2, routable_instances_count: 5, healthy_instances_count: nil, unhealthy_instances_count: 0)
+        up_scaler = DeploymentUpdater::Actions::UpScaler.new(deployment, logger, 10, summary)
+        expect(up_scaler.can_scale?).to be false
+      end
+
+      it 'returns false if routable instances already exceeds target' do
+        skip('Might need to restructure the scale action if we do this')
+        summary = double(starting_instances_count: 0, routable_instances_count: 5, healthy_instances_count: nil, unhealthy_instances_count: 0)
+        up_scaler = DeploymentUpdater::Actions::UpScaler.new(deployment, logger, 4, summary)
+        expect(up_scaler.can_scale?).to be false
+      end
+    end
+
+    describe '#finished_scaling?' do
+      let(:current_instance_count) { 5 }
+
+      it 'returns true if current instances have reached the target number of instances' do
+        summary = double(starting_instances_count: 0, routable_instances_count: 5, healthy_instances_count: 5, unhealthy_instances_count: 0)
+        up_scaler = DeploymentUpdater::Actions::UpScaler.new(deployment, logger, 5, summary)
+        expect(up_scaler.finished_scaling?).to be true
+      end
+
+      it 'returns false if current instances have not reached the target number of instances' do
+        summary = double(starting_instances_count: 0, routable_instances_count: 5, healthy_instances_count: 5, unhealthy_instances_count: 0)
+        up_scaler = DeploymentUpdater::Actions::UpScaler.new(deployment, logger, 6, summary)
+        expect(up_scaler.finished_scaling?).to be false
+      end
+
+      # TODO: Currently, deployments transition to COMPLETED even if there are some instances that have not started and become routable
```

Comment:
I was able to reproduce this by adding this to Dora's healthcheck endpoint:
```
 if ENV['CF_INSTANCE_INDEX'].to_i==3
     status 500
     return
 end

```

Given 6 instances and max of flight=2, the deployment will complete before the instance has started/become routable.  Will fix in a future PR